File: apps/api/app/api/conversation_timeline.py
from fastapi import APIRouter, Depends
from pydantic import BaseModel

from app.core.tenant_context import get_current_tenant
from app.db.conversation_timeline_repository import (
    create_internal_note,
    create_timeline_event,
    list_internal_notes,
    list_timeline_events,
)


router = APIRouter()


class CreateNoteRequest(BaseModel):
    note: str
    manager_id: str | None = None
    manager_name: str | None = None


@router.get("/inbox/conversations/{phone}/notes")
def get_notes(
    phone: str,
    tenant=Depends(get_current_tenant),
):
    org_id = tenant["org_id"]

    return {
        "status": "ok",
        "notes": list_internal_notes(
            org_id=org_id,
            client_phone=phone,
        ),
    }


@router.post("/inbox/conversations/{phone}/notes")
def add_note(
    phone: str,
    body: CreateNoteRequest,
    tenant=Depends(get_current_tenant),
):
    org_id = tenant["org_id"]

    note = create_internal_note(
        org_id=org_id,
        client_phone=phone,
        note=body.note,
        manager_id=body.manager_id,
        manager_name=body.manager_name,
    )

    create_timeline_event(
        org_id=org_id,
        client_phone=phone,
        event_type="NOTE_CREATED",
        event_title="Note interne ajoutee",
        event_payload={
            "note_id": str(note["id"]),
            "note": body.note,
        },
        created_by_id=body.manager_id,
        created_by_name=body.manager_name,
    )

    return {
        "status": "ok",
        "note": note,
    }


@router.get("/inbox/conversations/{phone}/timeline")
def get_timeline(
    phone: str,
    tenant=Depends(get_current_tenant),
):
    org_id = tenant["org_id"]

    return {
        "status": "ok",
        "events": list_timeline_events(
            org_id=org_id,
            client_phone=phone,
        ),
    }

File: apps/web/dashboard/services/conversation-timeline.ts
import { api } from "@/services/api";
import type {
  ConversationInternalNote,
  ConversationTimelineEvent,
} from "@/types/inbox";

export async function getConversationNotes(
  phone: string
): Promise<ConversationInternalNote[]> {
  const response = await api.get(
    `/inbox/conversations/${encodeURIComponent(phone)}/notes`
  );

  return response.data.notes;
}

export async function createConversationNote(
  phone: string,
  data: {
    note: string;
    manager_id?: string | null;
    manager_name?: string | null;
  }
): Promise<ConversationInternalNote> {
  const response = await api.post(
    `/inbox/conversations/${encodeURIComponent(phone)}/notes`,
    data
  );

  return response.data.note;
}

export async function getConversationTimeline(
  phone: string
): Promise<ConversationTimelineEvent[]> {
  const response = await api.get(
    `/inbox/conversations/${encodeURIComponent(phone)}/timeline`
  );

  return response.data.events;
}
